src/pip/_internal/utils/plugins.py

```diff
@@ -0,0 +1,72 @@
+import contextlib
+import logging
+from pathlib import Path
+from typing import Iterator, List
+
+from pip._vendor.pygments.plugin import iter_entry_points
+
+from pip._internal.models.plugin import DistInspectorPlugin, Plugin, plugin_from_module
+
+logger = logging.getLogger(__name__)
+
+_loaded_plugins: List[Plugin] = []
+for entrypoint in iter_entry_points(group_name="pip.plugins"):
+    try:
+        module = entrypoint.load()
+    except ModuleNotFoundError:
+        logger.warning("Tried to load plugin %s but failed", entrypoint.name)
+        continue
+    plugin = plugin_from_module(entrypoint.name, module)
+    if plugin is not None:
+        _loaded_plugins.append(plugin)
```

Comment:
I'm not a great fan of running complex code at import time. Couldn't we have a `load_plugins()` function that gets called at a suitable point in pip's initialisation process? Or better still, load plugins on demand, because we don't want to impact performance loading plugins we don't need.

Reply:
Good point! I moved it into a `load_plugins` and called right before `command.main(cmd_args)`.

For loading plugins on demand, maybe we can change `utils.plugins.plugin_pre_download_hook` and `utils.pugins.plugin_pre_extract_hook` to check if plugins have been loaded, and if not to do so. 
Then the plugins should only be loaded when the first hook is called (and never if the code path does not include any hooks).
WDYT?

Reply:
That'll do for now, but let's keep on-demand loading as an option, and do some proper testing at some point. I'm very aware that `uv` has demonstrated that performance is an important characteristic in an installer, and I want to make sure we avoid making pip slower and slower as we add features.